### PR TITLE
regression-report.yml should be independent of checkstyle/contribution and diff.groovy

### LINKED ISSUE #15665: regression-report.yml should be independent of checkstyle/contribution and diff.groovy
We currently rely on the diff.groovy, and groovy being a scripting language, can be slightly uncerstain. Hence, we start to use the java code mimicking the same functionality of `diff.groovy`. We have the java code ready and a part of `checkstyle/test-configs `completed as a part of https://github.com/checkstyle/test-configs/issues/105. 

Also, since diff.groovy is a part of checkstyle/contribution, we move away from it and rely on `checkstyle/test-configs`, which has the `DiffTool.java,` a migration of `diff.groovy`. And, We use the patch-diff-tool jar, and a new github action for posting comments back to PR. This makes us independnet of the checkstyle/contribution. It is a work in progress in https://github.com/checkstyle/test-configs/issues/104.

### PR DESCRIPTION
**Pull request by relentless-pursuit**
Part #15665

Moved to follow up PR:
Update for comments action 
Defect on new module mode

### CODE REVIEW

**Comment by romani** [comment]
@relentless-pursuit , please share links to tests that you did

**Comment by relentless-pursuit** [comment]
https://github.com/checkstyle-gsoc/checkstyle/actions/runs/10894292262

Run started

**Comment by romani** [comment]
Please share testing of all 3 modes, just to be sure we executed all of them ones and they worked.

**Comment by relentless-pursuit** [comment]
https://github.com/checkstyle-gsoc/checkstyle/actions/runs/10894476027

Run started for input java file

**Comment by relentless-pursuit** [comment]
https://github.com/checkstyle-gsoc/checkstyle/actions/runs/10894491563

Run started for config in PR description with patch config

**Comment by relentless-pursuit** [comment]
Please follow the testing results from here, https://github.com/checkstyle-gsoc/checkstyle/pull/28#issuecomment-2354288326

I had to update the workflow in my own repo having the necessary AWS S3 credentials. Please ignore above run as they will not be able to publish the reports.

File: .github/workflows/regression-report.yml [hunk 1]
```diff
@@ -491,36 +511,45 @@ jobs:
...
           REPO="${{ github.workspace }}/.ci-temp/checkstyle"
           BASE_BRANCH="upstream/master"
           export MAVEN_OPTS="-Xmx5g"
+
+          # Run DiffTool.jar with appropriate arguments
           if [ -f new_module_config.xml ]; then
             echo "Generating report for new module"
-            groovy diff.groovy -r "$REPO" -p "$PR_BRANCH" -pc new_module_config.xml -m single\
-              -l project.properties -xm "-Dcheckstyle.failsOnError=false"\
+            java -jar difftool.jar -r "$REPO" -p "$PR_BRANCH" -c new_module_config.xml -m single \
```
Comment (romani) [inline .github/workflows/regression-report.yml:540]: Please keep `-pc` it is " patch config".

Parameters should not change. Do same for all others.

File: .github/workflows/regression-report.yml [hunk 2]
```diff
@@ -454,10 +454,30 @@ jobs:
 
       - name: Prepare environment
         run: |
-          mv .ci-temp/project.properties ./.ci-temp/contribution/checkstyle-tester/
-          mv .ci-temp/*.xml ./.ci-temp/contribution/checkstyle-tester/
+          mv .ci-temp/project.properties ./.ci-temp/test-configs/checkstyle-tester/
+          mv .ci-temp/*.xml ./.ci-temp/test-configs/checkstyle-tester/
           sudo apt install groovy
```
Comment (romani) [inline .github/workflows/regression-report.yml:459]: Please remove this line , to be sure we do not use groovy.
Please Rename step, as it's goal becomes different

File: .github/workflows/regression-report.yml [hunk 3]
```diff
@@ -454,10 +454,30 @@ jobs:
...
+          repository: "relentless-pursuit/test-configs"
+          tag: "diff-tool-1.0.0"
+          fileName: "difftool.jar"
+
+      - name: Download patch-diff-report-tool JAR
+        uses: robinraju/release-downloader@v1.8
+        with:
+          repository: "checkstyle/test-configs"
+          tag: "patch-diff-report-tool-0.1-SNAPSHOT"
+          fileName: "patch-diff-report-tool-0.1-SNAPSHOT-jar-with-dependencies.jar"
```
Comment (romani) [inline .github/workflows/regression-report.yml:473]: This should be downloading from contribution repository, as contribution repo host sources, so we will always release to artifacts of contribution repository

File: .github/workflows/regression-report.yml [hunk 4]
```diff
@@ -491,36 +511,45 @@ jobs:
...
+          if [ -f .ci-temp/test-configs/checkstyle-tester/patch_config.xml ]; then
+          cat .ci-temp/test-configs/checkstyle-tester/patch_config.xml
           else
           echo "patch_config.xml not found"
           fi
           echo "-----------------------------"
 
+      - name: Copy JAR files to working directory
+        run: |
+          cp "${{ github.workspace }}/difftool.jar" .ci-temp/test-configs/checkstyle-tester/
```
Comment (romani) [inline .github/workflows/regression-report.yml:523]: Instead of this extra copy , can we put files directly where they should be ?
https://github.com/checkstyle/checkstyle/blob/978b70cd08600571c7db4e8d83fcd1fbc301bb0b/.github/workflows/google-java-format.yml#L44

File: .github/workflows/regression-report.yml [hunk 5]
```diff
@@ -636,7 +665,9 @@ jobs:
           ./.ci/append-to-github-output.sh "message" "$(cat .ci-temp/message)"
 
       - name: Comment PR
-        uses: checkstyle/contribution/comment-action@master
+        uses: peter-evans/create-or-update-comment@v3
```
Comment (romani) [inline .github/workflows/regression-report.yml:668]: This is unrelated issue, 
Ok, we tested it in this PR, lets now move this change to separate PR with linkage to proper issue and merge it quickly.

File: .github/workflows/regression-report.yml [hunk 6]
```diff
@@ -115,11 +115,11 @@ jobs:
           git checkout -b ${{ steps.get_pr_details.outputs.branch }} \
                forked/${{ steps.get_pr_details.outputs.branch }}
 
-      - name: Checkout contribution repository
+      - name: Checkout test-configs repository
```
Comment (romani) [inline .github/workflows/regression-report.yml:118]: What for we need this repository checkout?

File: .github/workflows/regression-report.yml [hunk 7]
```diff
@@ -454,10 +454,30 @@ jobs:
 
       - name: Prepare environment
         run: |
-          mv .ci-temp/project.properties ./.ci-temp/contribution/checkstyle-tester/
-          mv .ci-temp/*.xml ./.ci-temp/contribution/checkstyle-tester/
+          mv .ci-temp/project.properties ./.ci-temp/test-configs/checkstyle-tester/
+          mv .ci-temp/*.xml ./.ci-temp/test-configs/checkstyle-tester/
```
Comment (romani) [inline .github/workflows/regression-report.yml:458]: Please remind me why we need this files

File: .github/workflows/regression-report.yml [hunk 8]
```diff
@@ -454,10 +454,30 @@ jobs:
...
           sudo apt install groovy
 
+      - name: Download DiffTool JAR
+        uses: robinraju/release-downloader@v1.8
+        with:
+          repository: "relentless-pursuit/test-configs"
+          tag: "diff-tool-1.0.0"
+          fileName: "difftool.jar"
```
Comment (romani) [inline .github/workflows/regression-report.yml:466]: Reminder for me:
Do release of this to org repository

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: .github/workflows/regression-report.yml (hunk 3 above)
Comment (romani) [reply .github/workflows/regression-report.yml:473]: release is moved to contribution repo https://github.com/checkstyle/contribution/releases/tag/patch-diff-report-tool-0.1-SNAPSHOT

File: .github/workflows/regression-report.yml (hunk 8 above)
Comment (romani) [reply .github/workflows/regression-report.yml:466]: done: https://github.com/checkstyle/test-configs/releases/tag/diff-java-tool-1.0-SNAPSHOT-all
compiled by jdk21

File: .github/workflows/regression-report.yml (hunk 7 above)
Comment (romani) [reply .github/workflows/regression-report.yml:458]: we have https://github.com/checkstyle/checkstyle/blob/f66c74cb841544b72719bc7cfaa33f7d82db7ab0/.github/workflows/regression-report.yml#L20

why we need "*.xml" files?

File: .github/workflows/regression-report.yml (hunk 6 above)
Comment (romani) [reply .github/workflows/regression-report.yml:118]: might be same resolved by discussion at https://github.com/checkstyle/checkstyle/pull/15666/files#r1763166880

File: .github/workflows/regression-report.yml (hunk 7 above)
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml:458]: in generate_report, we cd to checkstyle-tester and inside that we access the project.properties and config.xml. But, this is not related to default project properties per say. I mean say we have both the files, we move it to checkstyle-tester and execute the java jar from there.

File: .github/workflows/regression-report.yml (hunk 2 above)
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: done

File: .github/workflows/regression-report.yml (hunk 3 above)
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml:468]: updated

File: .github/workflows/regression-report.yml (hunk 4 above)
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: done

File: .github/workflows/regression-report.yml (hunk 5 above)
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml:666]: okay, the issue i had created aimed at removing contribution entirely, wouldn't this qualify a part of the issue?

File: .github/workflows/regression-report.yml (hunk 6 above)
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml:118]: we need the checkout because we need checkstyle-tester folder inside test-configs where java jar would find the patch diff tool jar.

File: .github/workflows/regression-report.yml (hunk 7 above)
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: done

File: .github/workflows/regression-report.yml (hunk 8 above)
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: updated

File: .github/workflows/regression-report.yml (hunk 6 above)
Comment (romani) [reply .github/workflows/regression-report.yml:118]: Jar can we located at any place.
Lets add new parameter to diff tool to specify path to jar.

Comment (romani) [reply .github/workflows/regression-report.yml:118]: Or download jar to path we diff tool expects

Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml:118]: > Jar can we located at any place. Lets add new parameter to diff tool to specify path to jar.

parameter for path diff too jar?

Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml:118]: > Or download jar to path we diff tool expects

yes, we are downloading that. but, won't we require the cd to move to that place for java jar to execute it?

Comment (romani) [reply .github/workflows/regression-report.yml:118]: Why we need checkout of whole repository?

File: .github/workflows/regression-report.yml (hunk 1 above)
Comment (romani) [reply .github/workflows/regression-report.yml]: When new module is created, it does not exist in master branch, so execution of checkstyle (master branch) on config with new Check will fail with exception that such module does not exists.
when we do `-pc` we make it clear that execution on master should be skipped and we should execute only patch (PR branch) checkstyle that do have this new module.

File: .github/workflows/regression-report.yml (hunk 5 above)
Comment (romani) [reply .github/workflows/regression-report.yml:666]: the bigger you make PR, the more complicated to merge it, it is always better to split something to separate PR and reference the same issue.

File: .github/workflows/regression-report.yml (hunk 6 above)
Comment (romani) [reply .github/workflows/regression-report.yml:118]: please remove it, and show me error. We should not do checkout of repository just to create folders that our diff tool expects, we can create folders by "mkdir" if so required.

Comment (romani) [reply .github/workflows/regression-report.yml:118]: >yes, we are downloading that. but, won't we require the cd to move to that place for java jar to execute it?

download it with path specified, so all folders that required will be created during download action.

**Comment by romani** [comment]
@relentless-pursuit , please fix 2 mandatory review items: unnecessary checkout of test-configs and "-pc".

File: .github/workflows/regression-report.yml (hunk 6 above)
Comment (romani) [reply .github/workflows/regression-report.yml:118]: > parameter for path diff too jar?

in better world it should be `java -jar "$DIFFTOOL_JAR"  --patch-diff-report-tool-path .ci-temp/test-configs/checkstyle-tester/patch-diff-report-tool-0.1-SNAPSHOT-jar-with-dependencies.jar -r "$REPO" -p "$PR_BRANCH" -c "$NEW_MODULE_CONFIG" -m single ..... `

so all required elements are defined by parameters.
Or to be placed in same folder as `"$DIFFTOOL_JAR"` is located, so `--patch-diff-report-tool-path` would have default value as "." (current folder).

Comment (romani) [reply .github/workflows/regression-report.yml:118]: oh,
am I right that you need this repo just to get this folder and this file - https://github.com/checkstyle/test-configs/blob/main/checkstyle-tester/pom.xml
if yes, lets download it by curl.
In future we might move this file to checkstyle repo, as this config is part of workflow configuration files, or java-diff-tool will keep it as resource and create this file during execution, so will not need to deal with it at all.

File: .github/workflows/regression-report.yml (hunk 1 above)
Comment (romani) [reply .github/workflows/regression-report.yml]: if you postpone this to separate issue https://github.com/checkstyle/checkstyle/issues/15669
ok, please put comment below `echo "Generating report for new module"` 
`until #15669 this mode is not working`

File: .github/workflows/regression-report.yml (hunk 6 above)
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: done

File: .github/workflows/regression-report.yml (hunk 1 above)
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: done, i will introduce the functionality in subsequent PRs.

File: .github/workflows/regression-report.yml (hunk 5 above)
Comment (relentless-pursuit) [reply .github/workflows/regression-report.yml]: removed, done.

File: .github/workflows/regression-report.yml (hunk 6 above)
Comment (romani) [reply .github/workflows/regression-report.yml]: Improvement moved to later time https://github.com/checkstyle/test-configs/issues/190

**Comment by relentless-pursuit** [comment]
from config generation: https://github.com/checkstyle-gsoc/checkstyle/pull/28#issuecomment-2359346000
from existing bundle path: https://github.com/checkstyle-gsoc/checkstyle/pull/28#issuecomment-2359334938
from inputJava: https://github.com/checkstyle-gsoc/checkstyle/pull/28#issuecomment-2359366082
